components/director/pkg/authenticator/authenticator.go

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 The Compass Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package authenticator
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+)
+
+// InitFromEnv loads authenticator configurations from environment if any exist
+func InitFromEnv(envPrefix string) ([]Config, error) {
```

Comment:
I think you might have mistaken the `envconfig` library that we're using.
The functionalities that you referenced are based on: `github.com/kelseyhightower/envconfig`
We're using this much less feature-rich library: `github.com/vrischmann/envconfig`

Reply:
yup, they have the exact same envconfig and default tags so I've confused them.

The vrrischmann one has some array support (but this would mean switching to an array and moving the authenticator name as a field of the authConfig struct). Depending on how much you rely on it being a map this can be tried out.
```
var conf struct {
    Shards []struct {
        Name string
        Port int
    } `envconfig:"default={foobar;localhost:2929};{barbaz;localhost:2828}"`
}
```

They also have something like the map Decoder example from kelseyhightower's version of the lib but leveraging Unmarshal 
https://github.com/vrischmann/envconfig/commit/940aedfa1b5c601fd0b27b3113cf2857662aa36f. It looks very similar just using unmarshal instead of decode.

Reply:
Actually we don't insist on the end-result type to be a map - so that's not a limitation - a slice is fine.
So I tried the slice variation of the library, but I couldn't get it to work - it never seemed to resolve anything in the config struct.
I then went on to try with a custom Unmarshaler, the concept is simple - given a string (the env variable), we should define how it would be parsed into a struct.

For example:

```go
type Authenticators []AuthnConfig

var conf struct {
    Authns Authenticators
}
```

This has the implication of the Authns env var (`APP_AUTHNS`) to be a string containing all authenticators configuration. For example:
```
APP_AUTHNS='[{"name":"uaa1", "scopePrefix":"uaa1!abc", "attributes": {"uniqueAttribute": { "key": "parent_attr.child_attr", "value": "UAA1" }, "tenant": { "key": "parent_attr.tenantid" }, "identity": { "key": "username" } }}, {"name":"uaa2", "scopePrefix":"uaa2!abc", "attributes": {"uniqueAttribute": { "key": "parent_attr.child_attr", "value": "UAA2" }, "tenant": { "key": "parent_attr.tenantid" }, "identity": { "key": "username" } }}]'
```

```go
func (a *Authenticators) Unmarshal(s string) error {
        // unmarshal `s` into `a`
	return nil
}
```

Seems simple enough, but it adds complexity to the helm charts, they'd have to change in the following manner:
```
 authenticators:
    uaa1:
      enabled: true
      gatewayHost: "compass-gateway-uaa1"
      tokenURL: "..."
      jwksURL: "..."
      config: '{"name":"uaa1", "scopePrefix":"uaa1!abc", "attributes": {"uniqueAttribute": { "key": "parent_attr.child_attr", "value": "UAA1" }, "tenant": { "key": "parent_attr.tenantid" }, "identity": { "key": "username" } }}'
    uaa2:
      enabled: true
      gatewayHost: "compass-gateway-uaa2"
      tokenURL: "..."
      jwksURL: "..."
      config: '{"name":"uaa2", "scopePrefix":"uaa2!abc", "attributes": {"uniqueAttribute": { "key": "parent_attr.child_attr", "value": "UAA2" }, "tenant": { "key": "parent_attr.tenantid" }, "identity": { "key": "username" } }}'
```

So far it might be okay, but the tricky part comes when the respective `authenticators.<authn>.config` properties need to be concatenated in the form `[{...},{...}]` so that they can be easily unmarshaled in the code. I think Helm does have some primitives for variable declaration and manipulation so in the Director deployment it would be necessary to loop all authenticators and declare some local Helm variable which is concatenated with all authenticator configs.

Another option would be to maintain a separate variable for the authenticator configs which contains all authenticators but this will be strange considering that the old structure has to continue to exist beside it (enabled, gatewayHost, tokenURL and jwksURL need to be structured and accessible individually by helm to inject in the respective resources, so we cannot put all authenticator configuration "under the same umrella".

Additionally such a change would sort introduce sort of double maintanence of the authenticator name as it is both in the structured path and also in the config json variable.

All things considered I prefer to have complexity in the code instead of on the helm chart side, so I'd rather have the env processing as is.